File: Exemplos/Ex3_Veiculos/Ex3.py
# PRODUTO ABSTRATO
class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrar_detalhes(self):
        print(f"Modelo: {self.modelo}")

# PRODUTOS CONCRETOS
class Carro(Veiculo): pass
class Moto(Veiculo): pass

# FABRICA ABSTRATA
class FabricaDeVeiculos:
    def criar_veiculo(self, modelo): pass

# FABRICAS CONCRETAS
class FabricaDeCarros(FabricaDeVeiculos):
    def criar_veiculo(self, modelo): return Carro(modelo)

class FabricaDeMotos(FabricaDeVeiculos):
    def criar_veiculo(self, modelo): return Moto(modelo)

# CLIENTE
if __name__ == "__main__":
    fabrica_carros = FabricaDeCarros()
    fabrica_motos = FabricaDeMotos()

    v1 = fabrica_carros.criar_veiculo("Sedan")
    v2 = fabrica_carros.criar_veiculo("Hatch")
    v3 = fabrica_motos.criar_veiculo("Esportiva")
    v4 = fabrica_motos.criar_veiculo("Custom")

    v1.mostrar_detalhes()
    v2.mostrar_detalhes()
    v3.mostrar_detalhes()
    v4.mostrar_detalhes()

File: Factory Method/Exemplos/Ex1_Veiculos/Ex3.js
/**
 * PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)
 * 
 * Este padrão é útil quando uma classe não pode antecipar a classe 
 * de objetos que ela precisa criar.
 */

// 1. PRODUTO ABSTRATO (Simulado com classe base)
class Veiculo {
    constructor(modelo) {
        this.modelo = modelo;
    }

    mostrarDetalhes() {
        console.log(`Veículo Modelo: ${this.modelo}`);
    }
}

// 2. PRODUTOS CONCRETOS
class Carro extends Veiculo {
    constructor(modelo) {
        super(modelo);
    }
}

class Moto extends Veiculo {
    constructor(modelo) {
        super(modelo);
    }
}

// 3. CRIADOR ABSTRATO
class FabricaDeVeiculos {
    // Factory Method
    criarVeiculo(modelo) {
        throw new Error("Este método deve ser sobrescrito pelas subclasses.");
    }
}

// 4. CRIADORES CONCRETOS
class FabricaDeCarros extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Carro(modelo);
    }
}

class FabricaDeMotos extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Moto(modelo);
    }
}

// 5. CLIENTE
const fabricaCarros = new FabricaDeCarros();
const fabricaMotos = new FabricaDeMotos();

const v1 = fabricaCarros.criarVeiculo("SUV");
const v2 = fabricaMotos.criarVeiculo("Scooter");

v1.mostrarDetalhes();
v2.mostrarDetalhes();
